File: src/main/java/com/dude/dms/ui/builder/tags/TagEditDialogBuilder.java
package com.dude.dms.ui.builder.tags;

import com.dude.dms.backend.data.Tag;
import com.dude.dms.backend.service.DocService;
import com.dude.dms.backend.service.TagService;
import com.dude.dms.ui.EntityEditListener;
import com.dude.dms.ui.builder.BuilderFactory;
import com.dude.dms.ui.components.dialogs.TagEditDialog;

public final class TagEditDialogBuilder {

    private final BuilderFactory builderFactory;

    private final Tag tag;

    private final TagService tagService;

    private final DocService docService;

    private EntityEditListener<Tag> editListener;

    TagEditDialogBuilder(BuilderFactory builderFactory, Tag tag, TagService tagService, DocService docService) {
        this.builderFactory = builderFactory;
        this.tag = tag;
        this.tagService = tagService;
        this.docService = docService;
    }

    public TagEditDialogBuilder withEditListener(EntityEditListener<Tag> editListener) {
        this.editListener = editListener;
        return this;
    }

    public TagEditDialog build() {
        TagEditDialog dialog = new TagEditDialog(builderFactory, tag, tagService, docService);
        dialog.setEditListener(editListener);
        return dialog;
    }
}

File: src/main/java/com/dude/dms/ui/builder/tags/TagEditDialogBuilder.kt
package com.dude.dms.ui.builder.tags

import com.dude.dms.backend.brain.EditEvent
import com.dude.dms.backend.data.Tag
import com.dude.dms.backend.service.DocService
import com.dude.dms.backend.service.TagService
import com.dude.dms.ui.builder.BuilderFactory
import com.dude.dms.ui.components.dialogs.TagEditDialog

class TagEditDialogBuilder(
        private val builderFactory: BuilderFactory,
        private val tag: Tag,
        private val tagService: TagService,
        private val docService: DocService,
        private var editEvent: EditEvent<Tag>? = null) {

    fun build() = TagEditDialog(builderFactory, tag, tagService, docService).also { it.editEvent = editEvent }
}